File: InformationRetrieval/3_Fagin/Fagin.java
import java.util.*;
import java.io.*;

public class Fagin {
	public static void main(String arg[]) throws Exception{
		// java Fagin C:\Users\cecin\Desktop\DataScience\DMT\hw01\output\output_cranESS_bm25_title.tsv C:\Users\cecin\Desktop\DataScience\DMT\hw01\output\output_cranESS_bm25_text.tsv C:\Users\cecin\Desktop\DataScience\DMT\hw01\output\output_Fagin2.tsv k
		if(arg.length != 4) {
			System.out.println("\nWrong number of arguments: given "+arg.length
					+" arguments, the correct number is 4.\n");
			System.out.println("Correct use:\n");
			System.out.println("java fagin output_title_file.tsv output_text_file.tsv"
					+ " output_fagin_file.tsv k");
			return;
		}
		String titlef = arg[0];
		String textf = arg[1];
		String outputf = arg[2];
		int k = Integer.valueOf(arg[3]);
		Fagin pr = new Fagin();
		/*
		String path = "C:\\Users\\cecin\\Desktop\\DataScience\\DMT\\hw01\\output\\";
		String titlef = path+"output_cranESS_bm25_title.tsv";
		String textf = path+"output_cranESS_bm25_text.tsv";
		String outputf = path+"output_Fagin2.tsv";
		*/
		Map<String, String> cran_title,cran_text;
		cran_title = pr.createQueryDic(titlef);
		cran_text = pr.createQueryDic(textf);
		pr.printFaginResults(cran_title,cran_text,k,outputf);
	}
	private Map<String, String> createQueryDic(String cranFile) throws Exception {
		Map<String, String> cranDic = new HashMap<String, String>();
		String row, Query_ID = "", Doc_ID, Score, key, Docs= "";
   		FileReader fr = new FileReader(cranFile);
   		BufferedReader br = new BufferedReader(fr);
   		row = br.readLine();
   		int n = 0;
   		String[] app;
   		while((row = br.readLine())!=null) {
   			app = row.split("\\t");
   			if(app.length>1) {
   				if(Query_ID.length()>0 && Integer.parseInt(Query_ID)==Integer.parseInt(app[0])){
   					Docs += app[1]+" ";
   					n++;
   				}
   				else{
   					if(n>0){
   						cranDic.put("Q"+Query_ID+"_rank", Docs);
   					}
   					n = 0;
   					Docs = app[1]+" ";
   				}
   				Query_ID = app[0];
   				Doc_ID = app[1];
   				Score = app[3];
   				key = 'Q'+Query_ID+'_'+Doc_ID;
   				cranDic.put(key, Score);
   			}
   		}
		cranDic.put("Q"+Query_ID+"_rank", Docs);
		cranDic.put("nQueries", Query_ID);
   		br.close();
   		return(cranDic);
	}
	
	private Map<Float,ArrayList<String>>  fagin(Map<String, String> cran_title, Map<String, String> cran_text, int q, int k) {
		Set<String> cdocs = new HashSet<String>();
		Map<Float,ArrayList<String>> fdocs = new TreeMap<Float,ArrayList<String>>(Collections.reverseOrder());
		String doc;
		int count = 0, row = 0;
		String[] title_docs;
		String[] text_docs;
		if(cran_title.get("Q"+q+"_rank")!=null) {
			title_docs = cran_title.get("Q"+q+"_rank").split(" ");
		}
		else {
			title_docs = new String[0];
		}
		if(cran_text.get("Q"+q+"_rank")!=null) {
			text_docs = cran_text.get("Q"+q+"_rank").split(" ");
			
		}
		else {
			text_docs = new String[0];
		}
		while(count<k && row<title_docs.length && row<text_docs.length) {
			doc = title_docs[row];
			if(cdocs.contains(doc)==false) {
				cdocs.add(doc);
			}
			else {
				count ++;
			}
			doc = text_docs[row];
			if(cdocs.contains(doc)==false) {
				cdocs.add(doc);
			}
			else {
				count ++;
			}
			row ++;
		}
		
		if(count<k) {
			if(row<title_docs.length) {
				while(row<title_docs.length) {
					doc = title_docs[row];
					cdocs.add(doc);
					row ++;
				}
			}
			else {
				while(row<text_docs.length) {
					doc = text_docs[row];
					cdocs.add(doc);
					row ++;
				}
			}
		}
		
		for (String key: cdocs) {
			float Score = 0;
			String s;
			s = cran_title.get("Q"+q+"_"+key);
			if(s!=null) {
				Score = Float.valueOf(s)*2; 
			}
			s = cran_text.get("Q"+q+"_"+key);
			if(s!=null) {
				Score += Float.valueOf(s);
			}
        	ArrayList<String> d = new ArrayList<String>();
        	if (fdocs.containsKey(Score)==false) {
        		d.add(key);
        		fdocs.put(Score, d);
        	}
        	else {
        		d = fdocs.get(Score);
        		d.add(key);
        		fdocs.put(Score, d);
        	}
		}
		return(fdocs);
	}

	private void printFaginResults(Map<String, String> cran_title, Map<String, String> cran_text, int k, String outputFile) {
		try {
			FileOutputStream tsv_writer = new FileOutputStream(outputFile);
			PrintStream writerow = new PrintStream(tsv_writer);
			writerow.println("Query_ID\tDoc_ID\tRank\tScore");
			for(int q=1; q<Integer.valueOf(cran_title.get("nQueries"))+1; q++) {
				Map<Float,ArrayList<String>> result = fagin(cran_title,cran_text,q,k);
				int r = 1;
				for(Float score: result.keySet()) {
					for(String doc: result.get(score)) {
						if(r<k+1) {
							writerow.print(String.valueOf(q)+"\t"+doc+"\t");
							writerow.print(String.valueOf(r)+"\t"+String.valueOf(score));
							writerow.println();
							r += 1;
						}
					}
				}
			}
			writerow.close();
		}
		catch (IOException e) {
			System.out.println("Error: " + e);
		    System.exit(1);
		}
		
	}
}

File: InformationRetrieval/3_Fagin/fagin.py
import csv
import pandas as pd
import numpy as np
import operator

# given a DataFrame it returns a dictonary of dictionaries: the first key is the number of the query, the second is the 
# DocID and its score for that query, in 'rank' finally is stored the rank for the Docs in an ordered list of DocsID.
def createQueryDic(cranFile):
    n_queries = max(cranFile["Query_ID"].values)
    cranDic = {}
    for q in range(1,n_queries+1):
        docs =  list(cranFile[cranFile["Query_ID"] == q]["Doc_ID"])
        docs = list(map(str,docs))
        score =  list(cranFile[cranFile["Query_ID"] == q]["Score"])
        cranDic[q] = {str(docs[i]): score[i] for i in range(len(score))}
        cranDic[q]['rank'] = docs
    return(cranDic)

# given the dictonaries for the results of the queries for title and text it computes the fagin's algorithm for the 
# query q, for k results.
def fagin(cran_title, cran_text, q, k):
    cdocs = [] # candidate docsID
    count = 0 # how many couples of docs we've find till row r 
    r = 0 # the row number (or the rank)
    
    while count<k and r<len(cran_title[q]['rank']) and r<len(cran_text[q]['rank']):
        doc = cran_title[q]['rank'][r]
        if doc not in cdocs:
            cdocs.append(doc)
        else:
            count += 1
        doc = cran_text[q]['rank'][r]
        if doc not in cdocs:
            cdocs.append(doc)
        else:
            count += 1
        r += 1
        
    if k!=count:
        cdocs = set(cran_title[q]['rank']).union(cran_text[q]['rank'])
    # now we have to compute the scores for the candidate documents and we store them into fdocs
    fdocs = {} # final scores for the documents
    for doc in cdocs:
        score = 0
        if doc in cran_title[q].keys():
            score += cran_title[q][doc]*2 # two times for the title
        if doc in cran_text[q].keys():
            score += cran_text[q][doc] # one time for the text
        fdocs[doc] = score
    # finally k docs with the higher score
    fdocs['rank'] = [i for i,j in sorted(fdocs.items(), key=operator.itemgetter(1), reverse = True)][:k]
    return(fdocs)


# it prints the results into a tsv file.
def printResults(output,outputFileName):
    with open(outputFileName, 'w') as tsv_file:
        tsv_writer = csv.writer(tsv_file, delimiter = '\t')
        tsv_writer.writerow(['Query_ID', 'Doc_ID', 'Rank', 'Score']) # first row
        for query in sorted(output.keys()): # for each query
            r = 1
            for doc in output[query]['rank']: # the ordered docsID
                tsv_writer.writerow([str(query), doc, str(r), str(output[query][doc])])
                r += 1


#############################################################################################################

import sys
try:
    titlef = sys.argv[1]
    textf = sys.argv[2]
    outputf = sys.argv[3]
    k = int(sys.argv[4])
except:
    sys.exit(('\nWrong number of arguments: given '+str(len(sys.argv)-1)+' arguments, the correct number is 4.\n\n\n' 
              'Correct use:\n\n'
              'python fagin.py output_title_file.tsv output_text_file.tsv output_fagin_file.tsv k'))

# we import the input files into DataFrames 
cran_titlef = pd.read_csv(titlef,delimiter="\t")
cran_textf = pd.read_csv(textf,delimiter="\t")

# and then we create query dictionaries
cran_title = createQueryDic(cran_titlef)
cran_text = createQueryDic(cran_textf)


output = {}
# we compute the fagin's algorithm for each query
nquery = max(cran_title.keys())
for q in range(1,nquery+1):
    output[q] = fagin(cran_title, cran_text, q, k)

# and finally we save the results into the output file
printResults(output,outputf)



# python fagin.py C:\Users\cecin\Desktop\DataScience\DMT\hw01\output\output_cranESS_bm25_title.tsv C:\Users\cecin\Desktop\DataScience\DMT\hw01\output\output_cranESS_bm25_text.tsv C:\Users\cecin\Desktop\DataScience\DMT\hw01\output\output_Fagin.tsv 20


# java Fagin C:\Users\cecin\Desktop\DataScience\DMT\hw01\output\output_cranESS_bm25_title.tsv C:\Users\cecin\Desktop\DataScience\DMT\hw01\output\output_cranESS_bm25_text.tsv C:\Users\cecin\Desktop\DataScience\DMT\hw01\output\output_Fagin2.tsv 20